Sprint 17

AlexanderKorytin opened the pull request.

AlexanderKorytin committed:
в разметку MediaActivity добавлен контейнер ScrollView для нормального отображения на маленьких экранах

AlexanderKorytin committed:
спринт 17
- реализован di Koin для пакета search

AlexanderKorytin committed:
спринт 17
- реализован di Koin для пакета player

AlexanderKorytin committed:
спринт 17
- реализован di Koin для пакетов settings и sharing
- выполнет автоматический refotmat code

AlexanderKorytin committed:
спринт 17
- в конструктор SettingsRepositoryImpl добавлен объект App
- доработан module settingsDomainModule

kgornostaeva commented on .idea/deploymentTargetDropDown.xml:
```diff
@@ -12,6 +12,6 @@
         </deviceKey>
```
🍏 Добавь всю папку /.idea в файл  .gitignore, чтобы она не подгружалась в Pull request.
Вот так:
 /.idea

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/Util/App.kt:
```diff
@@ -2,18 +2,45 @@ package com.example.playlistmaker.Util
 
```
🍏 В Koltin принято именовать пакеты с маленькой буквы:
https://kotlinlang.org/docs/coding-conventions.html#naming-rules

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/Util/Creator.kt:
```diff
@@ -28,52 +11,5 @@ import com.example.playlistmaker.sharing.domain.api.SharingInteractor
 import com.example.playlistmaker.sharing.domain.impl.SharingInteractorImpl
 
 object Creator {
```
🍏 Creator уже можно полностью удалить, он больше не пригодится

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/player/di/PlayerDomainModule.kt:
```diff
@@ -0,0 +1,22 @@
+        MediaPlayerRepositoryImpl(mediaPlayer = get())
+    }
+
+    factory<MediaPlayerInteractor> {
```
🍏 В Koin появилиь удобные аналоги  single и factory: singleOf и factoryOf. Возможно, тебе понравиится их использовать: https://insert-koin.io/docs/reference/koin-core/dsl-update/

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/player/di/PlayerDomainModule.kt:
```diff
@@ -0,0 +1,22 @@
+        MediaPlayer()
+    }
+
+    factory<MediaPlayerRepository> {
```
🍏 Репозитории и интеракторы в большинстве случаев лучше делать синглтонами (с помощью метода single), чтобы сохраняемое ими состояние данных было одинаково во всех местах использования.

AlexanderKorytin replied:
Cделать single не получилось, т.к. при вызове метода onCleared() у MediaPlayerViewModel вызывается release() у созданного ранее в Koin экземпляра MediaPlayer.
   При повторном вызове MediaActivity - репозитория, если он single, не пересоздается, а значит MediaPlayerRepository начинает взаимодействовать с уничтоженным экземпляром MediaPlayer.
Видимо я что-то не так делаю

kgornostaeva commented on app/src/main/res/layout/activity_media.xml:
```diff
@@ -106,7 +111,7 @@
         android:layout_height="wrap_content"
         android:background="?attr/colorPrimary"
-        android:layout_marginTop="30dp"
+        android:layout_marginTop="38.33dp"
```
🍏 Какое-то магическое число) Если придется намного корректировать верстку, то будет непонятно, откуда оно взялось. Лучше использовать более стандартные отступы. Я вижу, что на макете отступ для этой кнопки составляет 30dp, размеры изображения - 100dp. Возможно, в проект сейчас скачан компонент, меньше по размерам, проверь еще раз и скачай избражение, у которого сразу есть дополнительный отступ.

https://www.figma.com/file/k9z167lAABCA7wt8gONM14/Playlist-Maker-(YP)?type=design&node-id=88-7009&mode=design&t=sZ1Cf73gWgCOrEzO-0

В работе мы тоже иногда сталкиваемся с тем, что дизайнер указал не те размеры или приложил некорректную картинку, которая при скачивании отображается неправильно. В этих случаях нужно идти к дизайнеру, чтобы он сделал нормально)

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/settings/ui/SettingsActivity.kt:
```diff
@@ -8,19 +8,17 @@ import com.example.playlistmaker.Util.setOnClickListenerWithViber
 import com.example.playlistmaker.Util.setVibe
```
🍏 Остались неиспользуемые импорты (и в других файлах тоже). Для форматирования кода и удаления неиспользуемых импортов применяй следующие сочетания клавиш:
Windows: Control+Alt+L
MacOS: Command+Option+L
Больше горячих клавиш можно найти здесь: https://developer.android.com/studio/intro/keyboard-shortcuts

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/settings/di/SettingsDomainModule.kt:
```diff
@@ -0,0 +1,21 @@
+import org.koin.dsl.module
+
+val settingsDomainModule = module {
+    single<App> { App() }
+
+    factory<SettingsRepository> {
+        SettingsRepositoryImpl(switchPreferences = get(), app = get())
+    }
```
🍏 Конутруктор для App обычно вызывает система, поэтому нам не требуется это делать. Koin имеет доступ к экземпляру нашего application, плюс при инициализации Koin ты передаешь ему ссылку на application. Поэтому достаточно сделать так:
```suggestion
    factory<SettingsRepository> {
        SettingsRepositoryImpl(switchPreferences = get(), app = androidApplication() as App)
    }
```

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/search/domain/impl/SearchTracksInteractorImpl.kt:
```diff
@@ -5,11 +5,11 @@ import com.example.playlistmaker.search.domain.api.TracksRepository
 import com.example.playlistmaker.search.domain.models.Track
 
-class SearchTracksInteractorImpl(private val repositiry: TracksRepository) :
+class SearchTracksInteractorImpl(private val trackRepositiry: TracksRepository) :
```
🍏 Опечатка в слове trackRepository

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/player/domain/api/MediaPlayerRepository.kt:
```diff
@@ -1,12 +1,15 @@
 package com.example.playlistmaker.player.domain.api
 
+import com.example.playlistmaker.player.data.dto.TrackUrl
```
⚠️ В слое domain не должно быть зависимостей от слоя data

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/player/domain/impl/MediaPlayerInteractorImpl.kt:
```diff
@@ -1,12 +1,17 @@
 package com.example.playlistmaker.player.domain.impl
 
+import com.example.playlistmaker.player.data.dto.TrackUrl
```
⚠️ В слое domain не должно быть зависимостей от слоя data

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/search/data/network/SearchHistoryRepositoryImpl.kt:
```diff
@@ -1,29 +1,27 @@
 
-class SearchHistoryRepositoryImpl(context: Context) :
+class SearchHistoryRepositoryImpl(
+    private val context: Context,
```
🍏 Контекст не используется, можно удалить

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/search/domain/impl/SearchTracksInteractorImpl.kt:
```diff
@@ -5,11 +5,11 @@ import com.example.playlistmaker.search.domain.api.TracksRepository
+class SearchTracksInteractorImpl(private val trackRepositiry: TracksRepository) :
     SearchTracksInteractor {
 
     override fun getMusic(term: String, consumer: Consumer<List<Track>>) {
```
🍏 Лучше называть аргументы так же, как они называются в интерфейсе или суперклассе, чтобы не возникало конфликтов:
```suggestion
    override fun getMusic(expression: String, consumer: Consumer<List<Track>>) {
```

kgornostaeva commented:
В результатах поиска картинки отображаются в разном размере, а должны быть одинаковыми квадратными. Используй CenterCrop, чтобы добиться нужного результата. При этом у Glide есть особенность: для того, чтобы все необходимые трансформации применились одновременно, нужно их указать в методе transform. Если отдельно прописывать .centerCrop().transform(RoundedCorners(..)) , то transform затрет centerCrop. 
Попробуй вот так: .transform(CenterCrop(), RoundedCorners(radiusIconTrackPx))

<img width="241" alt="Снимок экрана 2023-09-24 в 12 07 14" src="https://github.com/AlexanderKorytin/PlaylistMaker/assets/48408134/4ed4e30e-149a-49b4-96f6-96cf38d35a8e">

AlexanderKorytin committed:
спринт 17
- исправлены критические замечания ревью
- учтены все рекомендации кроме одного

AlexanderKorytin committed:
спринт 17
- исправлена ошибка создания экземпляра класса Gson() в FindActivity

AlexanderKorytin committed:
спринт 17
- util rename -> app

AlexanderKorytin committed:
спринт 17
- reformat code automatic

AlexanderKorytin committed:
написана функция возвращающая playerState

AlexanderKorytin committed:
спринт 17
- удален single получения App

AlexanderKorytin committed:
изменен экран медиа плеера - кнопка назад теперь всегда остается вверху и не уходит за экран при прокручивании экрана на меленьких смартфонах
